File: nim/day8.nim
import sequtils, strformat, tables, sets, tools

type 
    TI = tuple[x, y: int]
    HTSTI = Table[TI, seq[TI]]
    HCSTI = Table[char, seq[TI]]
    SC = seq[char]
    SSC = seq[SC]

const 
    width = 50
    height = 50

proc readInput(): (HCSTI, SSC) =
  let file = open("input/day8.txt", fmRead);
  defer: file.close()

  for i in file.lines():
    result[1] &= i.toSeq()

  for r_ind, r_val in result[1]:
    for c_ind, c_val in r_val:
      if c_val != '.':
        discard result[0].hasKeyOrPut(c_val, @[])
        result[0][c_val] &= (r_ind, c_ind)
         
func antSlopes(ants: HCSTI): HTSTI =
  for value in ants.values():
    for i in 0..<value.len() - 1:
      for j in i+1..<value.len():
        discard result.hasKeyOrPut(value[i], @[])
        result[value[i]] &= value[i] - value[j]

func solution(mat: SSC, slopes: HTSTI): int =
  var ttl = initHashSet[string]()
  for key, value in slopes:
    for i in value:
      var 
        symb = fetchVal(mat, key)
        pos = key + i
        neg = key - (i * 2)
      if inBounds(pos, width, height) and fetchVal(mat, pos) != symb:
        ttl.incl(pos.tupToStr())
      if inBounds(neg, width, height) and fetchVal(mat, neg) != symb:
        ttl.incl(neg.tupToStr())
        
  return ttl.len()
        
func solution2(slopes: HTSTI): int =
  var ttl = initHashSet[string]()
  for key, value in slopes:
    ttl.incl(key.tupToStr())
    for i in value:
      var val = key
      while true:
        val = val + i
        if inBounds(val, width, height):
          ttl.incl(val.tupToStr())
        else:
          val = key
          break
      while true:
        val = val - i
        if inBounds(val, width, height):
          ttl.incl(val.tupToStr())
        else: 
          break
  return ttl.len()
    
let 
  (ants, mat) = readInput()
  slopes = antSlopes(ants)
  pt1 = solution(mat, slopes)
  pt2 = solution2(slopes)
echo &"Solution 1: {pt1}\nSolution 2: {pt2}"

File: cpp/day8.cpp
#include "tools.h"
#include <iostream>
#include <fstream>
#include <unordered_map>
#include <unordered_set>
#include <cassert>

constexpr int ROWS = 50;
constexpr int COLS = 50;

typedef std::pair<int, int> pi; 
typedef std::array<std::array<char, COLS>, ROWS> aac; 
typedef std::unordered_map<char, std::vector<pi>> mcvpi;
typedef std::unordered_map<pi, std::vector<pi>, pair_hash> mpivpi;
typedef std::unordered_set<pi, pair_hash> spi;

struct ParseFile {aac arr; mcvpi mp;};

auto parseFile() -> ParseFile;
auto antSlopes(const mcvpi &ants) -> mpivpi;
auto solution(const aac &arr, const mpivpi &slopes) -> size_t;
auto solution2(const mpivpi &slopes) -> size_t;

auto main() -> int {
    auto [arr, ants] = parseFile();
    auto slopes = antSlopes(ants);
    auto sol1 = solution(arr, slopes);
    auto sol2 = solution2(slopes);

    std::cout << std::format("Part 1: {}\nPart 2: {}", sol1, sol2) << "\n";
}

auto parseFile() -> ParseFile {
    aac arr {{'.'}}; 
    mcvpi mp {{}};

    std::ifstream file ("input/day8.txt");
    assert(file);
    std::string line;

    size_t r_ind {0};
    while (std::getline(file, line)) {
        for (size_t c_ind {0}; c_ind < line.size(); ++c_ind) { 
            arr[r_ind][c_ind] = line[c_ind];
            if (mp.count(line[c_ind]) == 0) mp[line[c_ind]] = {};
            if (line[c_ind] != '.') mp[(char) line[c_ind]].emplace_back((int) r_ind, (int) c_ind);
        }
        ++r_ind;   
    }

    return {arr, mp};
}

auto antSlopes(const mcvpi &ants) -> mpivpi {
    mpivpi slopes {};

    for (const auto &vals : ants) {
        if (vals.second.size() < 2) continue;  
    
        for (size_t i {0}; i < vals.second.size() - 1; ++i) {
            for (size_t j {i + 1}; j < vals.second.size(); ++j) {
                slopes[vals.second[i]].emplace_back(vals.second[i] - vals.second[j]);
            }
        }
    }

    return slopes;
}

auto solution(const aac &arr, const mpivpi &slopes) -> size_t {
    spi sol1 {};

    for (const auto &vals : slopes) {
        for (const auto &i : vals.second) {
            auto vec = vals.first;
            auto symb = arr[i.first][i.second];
            auto pos = i + vals.first;
            auto neg = vec - (i * 2);

            if (inBounds(pos, ROWS, COLS) && arrValue(arr, pos) != symb) {
                sol1.emplace(pos);
            }

            if (inBounds(neg, ROWS, COLS) && arrValue(arr, neg) != symb) {
                sol1.emplace(neg);
            }
        }
    }

    return sol1.size();
}

auto solution2(const mpivpi &slopes) -> size_t {
    spi sol2 {};

    for (const auto &vals : slopes) {
        auto key = vals.first;
        auto value = vals.second;
        sol2.emplace(key);

        for (const auto &i : value) {
            auto val = key;
            while (true) {
                val = val + i;
                if (inBounds(val, ROWS, COLS)) {
                    sol2.emplace(val);
                } else {
                    val = key;
                    break;
                }
            }
            while (true) {
                val = val - i;
                if (inBounds(val, ROWS, COLS)) {
                    sol2.emplace(val);
                } else {
                    break;
                }   
            }
        }
    }
    
    return sol2.size();
}
